ClickHouse.Client/ADO/ClickHouseConnection.cs

```diff
@@ -277,7 +277,28 @@ public override async Task OpenAsync(CancellationToken cancellationToken)
     /// <param name="isCompressed">indicates whether "Content-Encoding: gzip" header should be added</param>
     /// <param name="token">Cancellation token</param>
     /// <returns>Task-wrapped HttpResponseMessage object</returns>
-    public async Task PostStreamAsync(string sql, Stream data, bool isCompressed, CancellationToken token)
+    public Task PostStreamAsync(string sql, Stream data, bool isCompressed, CancellationToken token)
+    {
+        var content = new StreamContent(data);
+        return PostStreamAsync(sql, content, isCompressed, token);
+    }
+
+    /// <summary>
+    /// Warning: implementation-specific API. Exposed to allow custom optimizations
+    /// May change in future versions
+    /// </summary>
+    /// <param name="sql">SQL query to add to URL, may be empty</param>
+    /// <param name="callback">Callback invoked to write to the stream. May contain SQL query at the beginning. May be gzip-compressed</param>
+    /// <param name="isCompressed">indicates whether "Content-Encoding: gzip" header should be added</param>
+    /// <param name="token">Cancellation token</param>
+    /// <returns>Task-wrapped HttpResponseMessage object</returns>
+    public Task PostStreamAsync(string sql, Func<Stream, CancellationToken, Task> callback, bool isCompressed, CancellationToken token)
```

Comment:
Same note about `async`/`await` - while this is a new function, it is worth maintaining consistency